tokio/src/task/local.rs

```diff
@@ -416,88 +389,58 @@ impl Scheduler {
-            Err(_) if std::thread::panicking() => return None,
-            Err(_) => panic!("mutex poisoned"),
-            Ok(lock) => lock,
-        };
-        lock.pop_front()
-    }
-
     fn tick(&self) {
         assert!(self.is_current());
         for _ in 0..MAX_TASKS_PER_TICK {
             let tick = self.tick.get().wrapping_add(1);
             self.tick.set(tick);
-            let task = match self.next_task(tick) {
+
+            let task = match unsafe {
+                // safety: we must be on the local thread to call this. The assertion
+                // the top of this method ensures that `tick` is only called locally.
+                self.queues.next_task(tick)
+            } {
                 Some(task) => task,
                 None => return,
             };
 
             if let Some(task) = task.run(&mut || Some(self.into())) {
                 unsafe {
-                    // we are on the local thread, so this is okay.
-                    self.schedule_local(task);
+                    // safety: we must be on the local thread to call this. The
+                    // the top of this method ensures that `tick` is only called locally.
+                    self.queues.push_local(task);
                 }
             }
         }
     }
-
-    fn drain_pending_drop(&self) {
-        for task in self.pending_drop.drain() {
-            unsafe {
-                (*self.tasks.get()).remove(&task);
-            }
-            drop(task);
-        }
-    }
-}
-
-impl fmt::Debug for Scheduler {
-    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
-        fmt.debug_struct("Scheduler { .. }").finish()
-    }
 }
 
 impl Drop for Scheduler {
     fn drop(&mut self) {
+        // Close the remote queue
+        self.queues.close_remote();
+
         // Drain all local tasks
-        while let Some(task) = self.next_local_task() {
+        while let Some(task) = unsafe { self.queues.next_local_task() } {
             task.shutdown();
         }
 
-        // Release owned tasks
         unsafe {
-            (*self.tasks.get()).shutdown();
-        }
+            // safety: these functions are unsafe to call outside of the local
+            // thread. Since the `Scheduler` type is not `Send` or `Sync`, we
+            // know it will be dropped only from the local thread.
 
-        self.drain_pending_drop();
+            // Release owned tasks
+            self.queues.shutdown();
 
-        // Wait until all tasks have been released.
-        // XXX: this is a busy loop, but we don't really have any way to park
-        // the thread here?
-        while unsafe { !(*self.tasks.get()).is_empty() } {
-            self.drain_pending_drop();
+            self.queues.drain_pending_drop();
+
+            // Wait until all tasks have been released.
+            // XXX: this is a busy loop, but we don't really have any way to park
+            // the thread here?
+            while self.queues.has_tasks_remaining() {
```

Comment:
tbh, i don't really know if this can happen. I believe the loop was copied from the threaded scheduler where a task could be "stolen" and running on a different thread and that thread has a "lock" on the task... anyway, not a big deal for now, but we should investigate it and better document the guarantees of `task` (it was initially not _really_ for reusing).